File: boj/Sort/BOJ_1517.cpp
// 백준 1517 버블 소트
// 21.02.08
#include <iostream>
#include <algorithm>
#include <queue>
#include <vector>
#include <string>
#include <cmath>
#define INF 2147000000

typedef long long ll;

using namespace std;

vector <ll> v;
ll ans = 0;

void mergeSort(int left, int right)
{
    if (left < right)
    {
        int mid = (left + right) / 2;
        mergeSort(left, mid);
        mergeSort(mid + 1, right);

        vector<ll> temp;
        int l = left;
        int r = mid + 1;
        while (l <= mid && r <= right)
        {
            if (v[l] > v[r])
            {
                temp.push_back(v[r++]);
                ans += mid - l + 1;
            }
            else
            {
                temp.push_back(v[l++]);
            }
        }
        while (l <= mid)
            temp.push_back(v[l++]);
        while (r <= right)
            temp.push_back(v[r++]);

        for (int i = 0; i < temp.size(); ++i)
            v[i + left] = temp[i];
        temp.clear();
    }
}

int main()
{
    int n; scanf("%d", &n);
    for (int i = 0; i < n; ++i)
    {
        ll a; scanf("%lld", &a);
        v.push_back(a);
    }

    mergeSort(0, n - 1);

    printf("%lld\n", ans);

    return 0;
}

File: boj/Sort/BOJ_1517.py
# 백준 1517 버블 소트
# 21.02.09

import sys
from functools import cmp_to_key
from collections import deque
sys.setrecursionlimit(10**6)


def mergeSort(left, right):
    global ans

    if left < right:
        mid = (left + right) // 2

        mergeSort(left, mid)
        mergeSort(mid + 1, right)
        
        temp = []
        l, r = left, mid + 1
        while l <= mid and r <= right:
            if nums[l] > nums[r]:
                temp.append(nums[r])
                r += 1
                ans += mid - l + 1
            else:
                temp.append(nums[l])
                l += 1
        
        while l <= mid:
            temp.append(nums[l])
            l += 1
        while r <= right:
            temp.append(nums[r])
            r += 1
        
        for i in range(len(temp)):
            nums[i + left] = temp[i]


n = int(input())

nums = [int(x) for x in sys.stdin.readline().split()]

ans = 0
mergeSort(0, n - 1)

print(ans)
